rustls-symcrypt/src/verify.rs

```diff
@@ -0,0 +1,356 @@
+use pkcs1::RsaPublicKey as AsnRsaPublicKey;
+use rustls::crypto::WebPkiSupportedAlgorithms;
+use rustls::pki_types::{AlgorithmIdentifier, InvalidSignature, SignatureVerificationAlgorithm};
+use rustls::SignatureScheme;
+use symcrypt::ecc::{CurveType, EcKey, EcKeyUsage};
+use symcrypt::hash::{sha256, sha384, sha512, HashAlgorithm};
+use symcrypt::rsa::{RsaKey, RsaKeyUsage};
+use webpki::alg_id::{self};
+
+/// Rsa signatures from the wire will come in the following ASN1 format:
+/// RSAPublicKey ::= SEQUENCE {
+///     modulus           INTEGER,  -- n
+///     publicExponent    INTEGER   -- e
+/// }
+fn extract_rsa_public_key(pub_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>), InvalidSignature> {
+    let key = AsnRsaPublicKey::try_from(pub_key).map_err(|_| InvalidSignature)?;
+
+    let modulus = key.modulus.as_bytes().to_vec();
+    let exponent = key.public_exponent.as_bytes().to_vec();
+
+    Ok((modulus, exponent))
+}
+
+// NistP256, NistP384, and NistP521 have a prepended Legacy byte that needs to be removed
+fn extract_ecc_public_key(pub_key: &[u8], curve_type: CurveType) -> Vec<u8> {
+    if curve_type == CurveType::Curve25519 {
+        // x25519 does not have the legacy byte prepended
+        return pub_key[..].to_vec();
+    }
+    // remove prepended 0x04 legacy byte
+    pub_key[1..].to_vec()
+}
+
+/// Ecc signatures from the wire will come in the following ASN1 format:
+/// ECDSASignature ::= SEQUENCE {
+///     r INTEGER,
+///     s INTEGER
+/// }
+/// SymCrypt expects a concatenated r+s with leading padding and leading 0's for both r and s to be removed
+fn extract_ecc_signature(signature: &[u8]) -> Result<Vec<u8>, InvalidSignature> {
+    // We pkcs1::RsaPublicKey because the underlying ASN1 format between an RSA public key and
+    // an ECC signgature is the same.
+    let signature = AsnRsaPublicKey::try_from(signature).map_err(|_| InvalidSignature)?;
+
+    // leading 0's are stripped when using as_bytes()
```

Comment:
>// leading 0's are stripped when using as_bytes() [](http://example.com/codeflow?start=4&length=49)

You might check with Mitch/Samuel. Will symcrypt ever generate an ECC signature where it can have a leading zero byte for r or s. This is different from the ASN.1 encoding where a leading 0 byte is added to r or s to treat it as being unsigned.

Reply:
Also, do other implementations do this trivial removal of leading zeroes.